agent/engine/docker_task_engine_test.go

```diff
@@ -814,6 +813,184 @@ func TestTaskTransitionWhenStopContainerTimesout(t *testing.T) {
+	mockTime.EXPECT().After(gomock.Any()).AnyTimes()
+	containerStoppingError := DockerContainerMetadata{
+		Error: &CannotStopContainerError{errors.New("Error stopping container")},
+	}
+	for _, container := range sleepTask.Containers {
+		gomock.InOrder(
+			imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes(),
+			client.EXPECT().PullImage(container.Image, nil).Return(DockerContainerMetadata{}),
+			imageManager.EXPECT().RecordContainerReference(container),
+			imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).Return(nil),
+			// Simulate successful create container
+			client.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Do(
+				func(x, y, z, timeout interface{}) {
+					go func() { eventStream <- dockerEvent(api.ContainerCreated) }()
+				}).Return(DockerContainerMetadata{DockerID: "containerId"}),
+			// Simulate successful start container
+			client.EXPECT().StartContainer("containerId", startContainerTimeout).Do(
+				func(id string, timeout time.Duration) {
+					go func() {
+						eventStream <- dockerEvent(api.ContainerRunning)
+					}()
+				}).Return(DockerContainerMetadata{DockerID: "containerId"}),
+			// StopContainer errors out a couple of times
+			client.EXPECT().StopContainer("containerId", gomock.Any()).Return(containerStoppingError).Times(2),
+			// Since task is not in steady state, progressContainers causes
+			// another invocation of StopContainer. Return the 'succeed' response,
+			// which should cause the task engine to stop invoking this again and
+			// transition the task to stopped.
+			client.EXPECT().StopContainer("containerId", gomock.Any()).Do(
+				func(id string, timeout time.Duration) {
+					go func() {
+						// Emit 'ContainerStopped' event to the container event stream
+						// This should cause the container and the task to transition
+						// to 'STOPPED'
+						eventStream <- dockerEvent(api.ContainerStopped)
+					}()
+				}).Return(DockerContainerMetadata{}),
+		)
+	}
+
+	err := taskEngine.Init()
+	assert.NoError(t, err, "Error getting event streams from engine")
+	taskEvents, contEvents := taskEngine.TaskEvents()
+
+	go taskEngine.AddTask(sleepTask)
+	// wait for task running
+	contEvent := <-contEvents
+	assert.Equal(t, contEvent.Status, api.ContainerRunning, "Expected container to be running")
+	*contEvent.SentStatus = api.ContainerRunning
+
+	taskEvent := <-taskEvents
+	assert.Equal(t, taskEvent.Status, api.TaskRunning, "Expected task to be running")
+	*taskEvent.SentStatus = api.TaskRunning
+
+	// Set the task desired status to be stopped and StopContainer will be called
+	updateSleepTask := *sleepTask
+	updateSleepTask.SetDesiredStatus(api.TaskStopped)
+	go taskEngine.AddTask(&updateSleepTask)
+
+	// StopContainer invocation should have caused it to stop eventually.
+	contEvent = <-contEvents
+	assert.Equal(t, contEvent.Status, api.ContainerStopped, "Expected container to be stopped")
```

Comment:
Should we check here the first two `StopContainer` didn't change/submit the task status to stopped?